### PR TITLE
Support conditionally raw value in `${value}`

### PR DESCRIPTION
**Pull request by AndiDog**
Similar to Django templates' [`safe`](https://docs.djangoproject.com/en/dev/ref/templates/builtins/#safe) filter or or Mako's [`n`](http://docs.makotemplates.org/en/latest/filtering.html#expression-filtering) filter, this provides a type `conditionally_raw_t<T>` whose value is escaped depending on how it was constructed. This way, you can safely use `${}` everywhere if this type is used.

My original use case was a helper function `formatTimestamp(time_t ts)` which then returns something like

``` cpp
auto formatTimestamp(time_t ts)
{
  return conditionally_raw_string::raw{"<span title=\"Timestamp " + std::to_string(ts) + "\">" + theFormattedDateAsSafeHtml + "</span>"};
}
```

as dynamic value inside a rendered table (simplified example):

``` cpp
%struct Row
%{
%  std::string label;
%  conditionally_raw_string value;
%};
%for (auto row : std::vector<Row>{
%  {"id", std::to_string(input.id)},
%  {"Sent to system X at", formatTimestamp(input.sentToOtherSystemTimestamp)},
%})
<tr>
  <th>${row.label}</th>
  <td>${row.value}</td>
</tr>
%}
```

### CODE REVIEW

**Comment by AndiDog** [comment]
Not sure if we can implicitly convert e.g. "const char*" to "conditionally_raw_string" in the current code, since I didn't test it against my real use case. In the changed example code, it requires extra braces in the initializer list. Maybe you find an improvement, if necessary.

File: include/kiste/raw_type.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    {
+    }
+
+    static auto raw(const T& t) -> conditionally_raw_t<T>
+    {
+      return conditionally_raw_t(t, true);
+    }
+
+    T t;
+    bool isRaw;
```
Comment (rbock) [inline include/kiste/raw_type.h]: This should be `_is_raw`.

**Comment by rbock** [comment]
Hi Andi!

I like that. 

I am a bit confused by the _conditionally_, though. I can immediately see uses for `always_raw`. What would be a use case for having rawness really optional?

Best,

Roland

**Comment by AndiDog** [comment]
I made some improvements: implicit construction from `const char*` if T convertible to string, and introduce `raw_t` and function `rawval` which creates it (you can now do `${kiste::rawval(expr)}`). Keeping `conditionally_raw_t` for use cases where both escaped/unescaped cases exist in one place (see my use case above). To support those cases, `conditionally_raw_t` can be constructed from `raw_t`, i.e. `conditionally_raw_string crs = kiste::rawval("<i>yay</i>");` works.

**Pull request merged**